File: benchmark/binary/x86-64/window/small/source/c/math_operations.c
#include <stdio.h>
#include <math.h>
#include <string.h>

// Basic math operations
int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

int lcm(int a, int b) {
    return (a / gcd(a, b)) * b;
}

// Prime number check
int is_prime(int n) {
    if (n <= 1) return 0;
    if (n <= 3) return 1;
    if (n % 2 == 0 || n % 3 == 0) return 0;
    
    for (int i = 5; i * i <= n; i += 6) {
        if (n % i == 0 || n % (i + 2) == 0) {
            return 0;
        }
    }
    return 1;
}

// Factorial
long factorial(int n) {
    if (n <= 1) return 1;
    return n * factorial(n - 1);
}

// Iterative factorial
long factorial_iter(int n) {
    long result = 1;
    for (int i = 2; i <= n; i++) {
        result *= i;
    }
    return result;
}

// Power function
long power(int base, int exp) {
    long result = 1;
    for (int i = 0; i < exp; i++) {
        result *= base;
    }
    return result;
}

// Modular exponentiation
long mod_power(long base, long exp, long mod) {
    long result = 1;
    base %= mod;
    
    while (exp > 0) {
        if (exp % 2 == 1) {
            result = (result * base) % mod;
        }
        exp = exp >> 1;
        base = (base * base) % mod;
    }
    
    return result;
}

// Sum of digits
int sum_of_digits(int n) {
    int sum = 0;
    n = n < 0 ? -n : n;
    
    while (n > 0) {
        sum += n % 10;
        n /= 10;
    }
    
    return sum;
}

// Count digits
int count_digits(int n) {
    if (n == 0) return 1;
    n = n < 0 ? -n : n;
    
    int count = 0;
    while (n > 0) {
        count++;
        n /= 10;
    }
    return count;
}

// Reverse integer
int reverse_integer(int n) {
    int reversed = 0;
    int is_negative = n < 0;
    n = is_negative ? -n : n;
    
    while (n > 0) {
        reversed = reversed * 10 + (n % 10);
        n /= 10;
    }
    
    return is_negative ? -reversed : reversed;
}

// Palindrome check
int is_palindrome_number(int n) {
    if (n < 0) return 0;
    return n == reverse_integer(n);
}

// Floating point operations
double circle_area(double radius) {
    return 3.14159 * radius * radius;
}

double circle_circumference(double radius) {
    return 2.0 * 3.14159 * radius;
}

double sphere_volume(double radius) {
    return (4.0 / 3.0) * 3.14159 * radius * radius * radius;
}

// Mean and variance
double calculate_mean(int *arr, int len) {
    int sum = 0;
    for (int i = 0; i < len; i++) {
        sum += arr[i];
    }
    return (double)sum / len;
}

double calculate_variance(int *arr, int len) {
    double mean = calculate_mean(arr, len);
    double variance = 0.0;
    
    for (int i = 0; i < len; i++) {
        double diff = arr[i] - mean;
        variance += diff * diff;
    }
    
    return variance / len;
}

// Matrix operations (2x2)
typedef struct {
    double a, b;
    double c, d;
} Matrix2x2;

Matrix2x2 matrix_multiply(Matrix2x2 m1, Matrix2x2 m2) {
    Matrix2x2 result;
    result.a = m1.a * m2.a + m1.b * m2.c;
    result.b = m1.a * m2.b + m1.b * m2.d;
    result.c = m1.c * m2.a + m1.d * m2.c;
    result.d = m1.c * m2.b + m1.d * m2.d;
    return result;
}

double matrix_determinant(Matrix2x2 m) {
    return m.a * m.d - m.b * m.c;
}

// Fibonacci with memoization
int fib_memo[100];

int fibonacci_memo(int n) {
    if (n <= 1) return n;
    if (fib_memo[n] != -1) return fib_memo[n];
    
    fib_memo[n] = fibonacci_memo(n - 1) + fibonacci_memo(n - 2);
    return fib_memo[n];
}

// Main function
int main() {
    // GCD and LCM
    int gcd_result = gcd(48, 18);
    int lcm_result = lcm(12, 18);
    
    // Prime check
    int prime_check = is_prime(17);
    
    // Factorial
    long fact = factorial_iter(5);
    
    // Power and modular exponentiation
    long pow_result = power(2, 10);
    long mod_pow_result = mod_power(2, 100, 1000000007);
    
    // Number operations
    int digit_sum = sum_of_digits(12345);
    int digit_count = count_digits(98765);
    int reversed = reverse_integer(12345);
    int is_palin = is_palindrome_number(121);
    
    // Circle operations
    double area = circle_area(5.0);
    double circumference = circle_circumference(5.0);
    double vol = sphere_volume(3.0);
    
    // Statistics
    int data[] = {10, 20, 30, 40, 50};
    double mean = calculate_mean(data, 5);
    double variance = calculate_variance(data, 5);
    
    // Matrix operations
    Matrix2x2 m1 = {1, 2, 3, 4};
    Matrix2x2 m2 = {5, 6, 7, 8};
    Matrix2x2 product = matrix_multiply(m1, m2);
    double det = matrix_determinant(m1);
    
    // Fibonacci memo
    memset(fib_memo, -1, sizeof(fib_memo));
    int fib_result = fibonacci_memo(20);
    
    printf("Math results: %d %d %ld %d\n", gcd_result, lcm_result, fact, fib_result);
    
    return 0;
}

File: benchmark/binary/x86-64/window/small/source/golang/math_operations.go
package main

import (
	"fmt"
	"math"
)

// Number theory
func gcd(a, b int) int {
	for b != 0 {
		temp := b
		b = a % b
		a = temp
	}
	return a
}

func lcm(a, b int) int {
	return (a / gcd(a, b)) * b
}

func isPrime(n int) bool {
	if n <= 1 {
		return false
	}
	if n <= 3 {
		return true
	}
	if n%2 == 0 || n%3 == 0 {
		return false
	}

	for i := 5; i*i <= n; i += 6 {
		if n%i == 0 || n%(i+2) == 0 {
			return false
		}
	}
	return true
}

// Factorial
func factorial(n int) int64 {
	if n <= 1 {
		return 1
	}
	return int64(n) * factorial(n-1)
}

func factorialIter(n int) int64 {
	result := int64(1)
	for i := 2; i <= n; i++ {
		result *= int64(i)
	}
	return result
}

// Power
func power(base, exp int) int64 {
	result := int64(1)
	for i := 0; i < exp; i++ {
		result *= int64(base)
	}
	return result
}

// Modular exponentiation
func modPower(base, exp, mod int64) int64 {
	result := int64(1)
	base = base % mod

	for exp > 0 {
		if exp%2 == 1 {
			result = (result * base) % mod
		}
		exp = exp >> 1
		base = (base * base) % mod
	}

	return result
}

// Digit operations
func sumOfDigits(n int) int {
	sum := 0
	if n < 0 {
		n = -n
	}

	for n > 0 {
		sum += n % 10
		n /= 10
	}

	return sum
}

func countDigits(n int) int {
	if n == 0 {
		return 1
	}
	if n < 0 {
		n = -n
	}

	count := 0
	for n > 0 {
		count++
		n /= 10
	}
	return count
}

func reverseInteger(n int) int {
	reversed := 0
	isNegative := n < 0
	if isNegative {
		n = -n
	}

	for n > 0 {
		reversed = reversed*10 + (n % 10)
		n /= 10
	}

	if isNegative {
		return -reversed
	}
	return reversed
}

func isPalindrome(n int) bool {
	if n < 0 {
		return false
	}
	return n == reverseInteger(n)
}

// Geometry
func circleArea(radius float64) float64 {
	return math.Pi * radius * radius
}

func circleCircumference(radius float64) float64 {
	return 2.0 * math.Pi * radius
}

func sphereVolume(radius float64) float64 {
	return (4.0 / 3.0) * math.Pi * radius * radius * radius
}

// Statistics
func calculateMean(arr []int) float64 {
	if len(arr) == 0 {
		return 0.0
	}
	sum := 0
	for _, v := range arr {
		sum += v
	}
	return float64(sum) / float64(len(arr))
}

func calculateVariance(arr []int) float64 {
	if len(arr) == 0 {
		return 0.0
	}

	mean := calculateMean(arr)
	variance := 0.0

	for _, v := range arr {
		diff := float64(v) - mean
		variance += diff * diff
	}

	return variance / float64(len(arr))
}

// Matrix2x2
type Matrix2x2 struct {
	a, b, c, d float64
}

// Multiply matrices
func (m1 *Matrix2x2) Multiply(m2 *Matrix2x2) *Matrix2x2 {
	return &Matrix2x2{
		a: m1.a*m2.a + m1.b*m2.c,
		b: m1.a*m2.b + m1.b*m2.d,
		c: m1.c*m2.a + m1.d*m2.c,
		d: m1.c*m2.b + m1.d*m2.d,
	}
}

// Determinant
func (m *Matrix2x2) Determinant() float64 {
	return m.a*m.d - m.b*m.c
}

// Fibonacci with memoization
func fibonacci(n int, cache map[int]int) int {
	if n <= 1 {
		return n
	}
	if val, ok := cache[n]; ok {
		return val
	}

	result := fibonacci(n-1, cache) + fibonacci(n-2, cache)
	cache[n] = result
	return result
}

func main() {
	// Test number theory
	gcdResult := gcd(48, 18)
	lcmResult := lcm(12, 18)
	isPrimeResult := isPrime(17)

	// Test factorial
	factResult := factorialIter(5)

	// Test power
	powResult := power(2, 10)
	modPowResult := modPower(2, 100, 1000000007)

	// Test digit operations
	digitSum := sumOfDigits(12345)
	digitCount := countDigits(98765)
	reversed := reverseInteger(12345)
	isPalinResult := isPalindrome(121)

	// Test geometry
	circleAreaResult := circleArea(5.0)
	circlCircumResult := circleCircumference(5.0)
	sphereVolResult := sphereVolume(3.0)

	// Test statistics
	data := []int{10, 20, 30, 40, 50}
	mean := calculateMean(data)
	variance := calculateVariance(data)

	// Test matrix operations
	m1 := &Matrix2x2{a: 1, b: 2, c: 3, d: 4}
	m2 := &Matrix2x2{a: 5, b: 6, c: 7, d: 8}
	product := m1.Multiply(m2)
	det := m1.Determinant()

	// Test Fibonacci with memoization
	cache := make(map[int]int)
	fibResult := fibonacci(15, cache)

	fmt.Printf("Math operations test: %d %d %d %d\n",
		gcdResult, lcmResult, int(factResult), digitSum)
	fmt.Printf("Powers: %d %d, Fibonacci: %d\n", powResult, modPowResult, fibResult)
	fmt.Printf("Geometry: area=%.2f, circumference=%.2f, volume=%.2f\n",
		circleAreaResult, circlCircumResult, sphereVolResult)
	fmt.Printf("Statistics: mean=%.2f, variance=%.2f\n", mean, variance)
	fmt.Printf("Matrix determinant: %.2f, Product a=%.2f\n", det, product.a)
	fmt.Printf("Palindrome: %v, Digits: %d, Prime: %v, Reversed: %d\n", isPalinResult, digitCount, isPrimeResult, reversed)
}
